## Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек
Выполнил студент гр. 9303 Павлов Дмитрий, вариант 17
## Цель работы
Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, привить навык использования систем компьютерной алгебры для реализации алгоритмов.
## Основные теоретические сведения
#### Бинарный метод
Бинарный метод - это один из алгоритмов быстрого возведения числа $x$ в некоторую степень $n\in \mathbb{N}$.

Алгоритм заключается в следующем: для начала числа $x$ представляется в двоичном виде, после чего из этого двоичного представления удаляется старший бит (единица). Далее в цикле просматриваются биты от старшего к младшему: если бит 0, то текущее число умножается на себя, если бит 1 - то число умножается на себя и на $x$.

Количество операций, требуемых для возведения числа $x$ в степень $n$, для данного метода равно $\lambda(n)+\nu(n)-1$, где $\lambda(n)=\lceil log_2(n) \rceil$, а $\nu(n)$ равно количеству единиц в двоичной записи числа $n$.

#### Метод множителей
Метод множителей это один из алгоритмов быстрого возведения числа $x$ в некоторую степень $n\in \mathbb{N}$.

Алгоритм заключается в следующем: для начала число $n$ раскладывается на множители ($n=i*j$), где $i$ - минимальный простой делитель числа $n$). Далее при помощи бинарного метода число &x& возводится в степень $i$, после чего полученный результат $x^i=y$ возводится в степень $j$: $y^j={(x^i)}^j=x^{i\cdot j}=x^n$. Если $n$ - простое число, то алгоритм сначала возводит число $x$ в степень $n-1$, а потом домножает на $x$, получая тем самым $x^n$.

#### Аддитивные цепочки
Аддитивная цепочка для некоторого числа $n\in \mathbb{N}$ - это последовательность натуральных чисел $\{a_i\}_{i=0}^m$, где $a_0 = 1$, $a_m = n$, в которой каждый последующий элемент является суммой каких-то двух предшествующих элементов.

$a_i = a_j + a_k, \forall i:1..m, k \leq j < i$
$l(n) = m$

Типы шагов в аддитивной цепочке:

Шаг $i$ называют удвоением, если $i - 1 = k = j$;

Шаг $i$ называют звездным, если $j = i - 1$, $k \in \{0, \dots, i-1\}$;

Шаг $i$ называют малым, если $\lambda(a_i)=\lambda(a_{i-1})$

Звездная цепочка - это аддитивная цепочка, в которая состоит только из звездных шагов.

#### Алгоритм Брауэра
Алгоритм Брауера вычисляет n-ую стпенень за $\lambda(n)+\frac{(1+o(1))\lambda(n)}{\lambda(\lambda(n))}$ операций.

Для некоторых $n, k$ Брауерские цепочки задаются в виде рекурентной формулы:
$$B_k(n) =\begin{cases}1, 2, 3, ..., 2^k-1\text{, если }n < 2^k \\ B_k(q), 2q, 4q, 8q, ..., 2^kq, n,\text{ если } n \geqslant 2^k\ \text{и } q = \lfloor\frac{n}{2^k}\rfloor \end{cases}$$

#### Алгоритм дробления вектора индексов
Алгоритм дробления вектора индексов позволяет найти минимальную звездную цепочку для некоторого числа $n \in \mathbb{N}$.

Рассмотрим вектор индексов $\{r_i\}_{i=1}^q \cup {\{{\rho}_j \}}_{j=q+1}^m$, где ${\rho}_j= \{x: 1 \leq x \leq j \}$, ${\{r_i\}}_{i=1}^q$ - фиксированная часть, ${\{{\rho}_j\}}_{j=q+1}^m$ - изменяющаяся часть.

Наибольшее значение $a_m$ достигается при векторе индексов ${\{r_i\}}_{i=1}^{q} \cup \{q+1,q+2,\dots,m\}$.

Наименьшее значение $a_m$ достигается при векторе индексов ${\{r_i\}}_{i=1}^{q}\cup\{1,1,\dots,1\}$.

$a_{max} = a_{q+1} \cdot {2}^{m-q}$

$a_{min} = a_{q+1}+m-q$

Алгоритм:

1) Во внешнем цикле рассматриваем аддитивные цепочки длины $m$ от значения $\underline{l}(n)=\lceil log_2(n) \rceil$ до $\bar{l}(n)=\lambda(n)+\nu(n)-1$, на каждой итерации выбираем $q$ ($1 \leq q \leq m-1$), пусть $q = \frac{m}{2}$

2) Далее перебираем все возможные фиксированные части вектора индексов $\{r_i\}_{i=1}^q$ ($q!$ вариантов), для каждой строим соответствующую ей звездную цепочку, находим $a_{max}$ и ${a}_{min}$

&emsp; а) Если $a_m = n$ - конец

&emsp; б) Если $n \notin [a_{min},a_{max}]$, то переходим к следующему набору $\{r_i\}_{i=1}^q$

&emsp; в) Если $n\in [a_{min},a_{max}]$, то перебираем все возможные изменяющиеся части вектора индексов ${\left \{{\rho}_j \right \}}_{ j=q+1}^m$ и находим $a_m$. Если $a_m=n$, то цепочка найдена. Если все возможные изменяющиеся части вектора индексов ${\{{\rho}_j\}}_{j=q+1}^m$ исчерпаны, то переходим к следующему набору $\{r_i\}_{i=1}^q$.

3) Если все наборы вектора индексов длины $m$ исчерпаны, то увеличиваем $m$ на 1.

#### Теорема Брауэра
Пусть $k < log_2(log_2(n))$ верно: $l(n) < (1+k^{-1}) * \lceil log_2(n) \rceil + 2^{k-1} - k + 2$

Тогда $k=\lambda(\lambda(n))-2\lambda(\lambda(\lambda(n)))$

*Следствие 1:* $\lim \limits_{n \to \infty} \frac{l(n)}{\lambda(n)} = 1$

*Следствие 2:* $\lambda(n)(1+\frac{1}{\lambda(\lambda(n))}+\frac{o(\lambda(\lambda(\lambda(n))))}{\lambda(\lambda(n))^2})$

#### Гипотеза Шольца-Брауэра
Пусть $l^*(n)$ - длина звёздной цепочки.

Тогда для любого $n \in \mathbb{N}$ верно: $l^*(2^n-1)\leq l^*(n)+n-1$

## Постановка задачи
Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной алгебры SageMath, провести анализ алгоритмов. Полученные результаты содержательно проинтерпретировать.
## Порядок выполнения работы
1. Применить бинарный метод и метод множителей для $x^n$, где $n\geq30$, для 2-3 значений n (значения n выбирается студентом самостоятельно). Сравнить количество операций для каждого метода, сделать выводы.
2. Реализовать алгоритм Брауэра (для нечётных вариантов) или алгоритм Яо (для чётных вариантов) для вычисления приближённых аддитивных цепочек для различных чисел при варьировании параметра k, сопоставить длины полученных аддитивных цепочек с минимальной аддитивной цепочкой для заданного числа. Сделать выводы.
3. Реализовать алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа. Протестировать алгоритм при $n>500$. Указать, сколько времени потребовалось на поиск цепочки и какая цепочка получилась. Сравнить с предыдущими методами, сделать выводы.
4. Проверить следствие 1 теоремы Брауэра для $n=1..200$ путём построения функции $l(n)$ и аппроксимирующей кривой, полученной с помощью метода наименьших квадратов. Сопоставить функции на одном графике, сделать выводы.
5. Проверить гипотезу Шольца–Брауэра для $1<n\leq10$ на алгоритме дробления вектора индексов. Сделать выводы.
6. *Дополнительное необязательное задание*: найти и/или предложить модификации алгоритмов и привести описание модификаций. Реализовать модифицированные алгоритмы и сравнить их мощность.

## Выполнение работы
*1) Бинарный метод и метод множителей*

In [5]:
def bin_meth(x, y):
    if y == 0:
        return 1
    if y == -1:
        return 1. / x
    p = bin_meth(x, y // 2)
    p *= p
    if y % 2:
        p *= x
    return p

def bin_oper_count(n):
    lambda_n = len(list(bin(n))[2:]) - 1
    nu_n = list(bin(n))[2:].count('1')
    return lambda_n + nu_n - 1

In [6]:
def factors(n):
    res = []
    d = 2
    while 1:
        if n % d == 0:
            res.append(d)
            res.append(n // d)
            break
        else:
            d += 1
    
    return res

def multiplier_meth(x, n):
    factors_n = factors(n)
    res = x
    for el in factors_n:
        res = bin_meth(res, el)
    return res

def mult_oper_count(n):
    factors_n = factors(n)
    res = 0
    res += bin_oper_count(factors_n[0]) + bin_oper_count(factors_n[1])
    return res

In [9]:
print("Введите число: ")
x = int(input())
print("Введите степень: ")
n = int(input())

print("\nБинарный метод:")
print(f"{x}^{n} =", bin_meth(x, n), "Число операций:", bin_oper_count(n))

print("\nМетод множителей:")
print(f"{x}^{n} =", multiplier_meth(x, n), "Число операций:", mult_oper_count(n))

Введите число: 
5
Введите степень: 
3

Бинарный метод:
5^3 = 125 Число операций: 2

Метод множителей:
5^3 = 125 Число операций: 2


Запишем результаты в таблицу:

| n | Бинарый метод | Метод множителей |
|:-:|:-:|:-:|
| 40 | 6 | 6 |
| 125 |11 | 9 |
| 255 | 14 | 11 |

По результам видно что бинарный метод и метод множителей работают примерно одинаково, но иногда метод множетелей работает быстрее.


*2) Алгоритм Брауэра*

In [1]:
def brouwer(n, k, chain):
    if(n < 2**k):
        for i in range(1, 2**k):
            chain.append(i)
        return    
    elif(n >=2**k):
        q = n//(2**k)
        q1=q
        brouwer(q, k, chain)
        for i in range(1, k+1):
            q*=2
            if(not(q1<2**(k-i))):
                chain.append(q)
        if n!= q:    
            chain.append(n)    

In [3]:
print("Введите число: ")
n = int(input())
for k in range(1,5):
    chain = []
    brouwer(n, k, chain)
    print("Число:", n,"k:", k, "\nЦепочка:", chain, "\nДлина цепочки:", len(chain), '\n')

Введите число: 
501
Число: 501 k: 1 
Цепочка: [1, 2, 3, 6, 7, 14, 15, 30, 31, 62, 124, 125, 250, 500, 501] 
Длина цепочки: 15 

Число: 501 k: 2 
Цепочка: [1, 2, 3, 4, 7, 14, 28, 31, 62, 124, 125, 250, 500, 501] 
Длина цепочки: 14 

Число: 501 k: 3 
Цепочка: [1, 2, 3, 4, 5, 6, 7, 14, 28, 56, 62, 124, 248, 496, 501] 
Длина цепочки: 15 

Число: 501 k: 4 
Цепочка: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 31, 62, 124, 248, 496, 501] 
Длина цепочки: 22 



#### Проведем тесты для различных значений n и k и сравним их с известными минимальными значенями длин цепочек:

| n | k | Цепочка | Длина цепочки | Минимальная длина |
|:-:|:-:|:-------:|:-------------:|:-----------------:|
| 35 | 2 | [1, 2, 3, 4, 8, 16, 32, 35] | 8 | 8 |
| 35 | 3 | [1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 35] | 11 | 8 |
| 35 | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 35] | 18 | 8 |
| 66 | 2 | [1, 2, 3, 4, 8, 16, 32, 64, 66] | 9 | 8 |
| 66 | 3 | [1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 64, 66] | 12 | 8 |
| 66 | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 64, 66] | 19 | 8 |
| 105 | 2 | [1, 2, 3, 4, 6, 12, 24, 26, 52, 104, 105] | 11 | 10 |
| 105 | 3 | [1, 2, 3, 4, 5, 6, 7, 8, 13, 26, 52, 104, 105] | 13 | 10 |
| 105 | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 24, 48, 96, 105] | 19 | 10 |

Можно заметить, что чем меньше значение k, тем длина цепочки приближается к мнимальной длине.

*3) Алгоритм дробления вектора индексов*

In [ ]:
import math
import time


def next_vector_r(vector, q):
    for i in range(q, 1, -1):
        if vector[i - 1] > 1:
            vector[i - 1] -= 1
            break
        if vector[i - 1] == 1:
            vector[i - 1] = i
    return vector


def next_vector_ro(vector, q, m):
    for i in range(m, q, -1):
        if vector[i - 1] > 1:
            vector[i - 1] -= 1
            break
        if vector[i - 1] == 1:
            vector[i - 1] = i
    return vector


def splitting_vector(n):
    low_l = int(math.log2(n)) + 1
    high_l = low_l - 1 + bin(n).count('1')
    for m in range(low_l, high_l + 1):
        vector = [x for x in range(1, m)]
        q = m // 2 - 1
        flag = 0
        while True:
            chain = [1]
            for i in vector:
                chain.append(chain[-1] + chain[i - 1])
            if chain[m - 1] == n:
                return chain
            if len(vector[:q]) == sum(vector[:q]):
                flag = 1
            a_min = chain[q] + m - q
            a_max = chain[q] * 2 ** (m - q)
            if n in range(a_min, a_max + 1):
                while True:
                    chain = [1]
                    vector = next_vector_ro(vector, q, m - 1)
                    for i in vector:
                        chain.append(chain[-1] + chain[i - 1])
                    if chain[m - 1] == n:
                        return chain
                    if len(vector[q:]) == sum(vector[q:]):
                        break
                vector = next_vector_r(vector, q)
            else:
                vector = next_vector_r(vector, q)
            if flag == 1:
                break
    return chain


start_time = time.time()
chain = splitting_vector(int(input()))
print(" %s seconds" % (time.time() - start_time))
print('chain = ', chain)

Проведем тесты алгоритма ДВИ при $n > 500$ и сравним полученные результаты с результатами работы алгоритма Брауэра (при $k=3$).

Сведем полученные данные в таблицу:

| n | Цепочка алгоритма Брауэра ($k = 2$) | Длина цепочки | Цепочка алгоритма ДВИ | Длина цепочки | Время работы алгоритма |
|:----:|:----:|:----:|:----:|:----:|:----:|
| 501 | [1, 2, 3, 4, 7, 14, 28, 31, 62, 124, 125, 250, 500, 501] | 14 | [1, 2, 4, 8, 16, 32, 64, 96, 100, 200, 400, 500, 501] | 13 | 92 сек.|
| 524 | [1, 2, 3, 4, 8, 16, 32, 64, 128, 131, 262, 524] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 520, 524]| 12 | 11 сек.|
| 571 | [1, 2, 3, 4, 8, 16, 32, 35, 70, 140, 142, 284, 568, 571] | 14 | [1, 2, 4, 8, 16, 32, 34, 35, 67, 134, 268, 536, 571]| 13 | 98 сек. |
| 638 | [1, 2, 3, 4, 8, 9, 18, 36, 39, 78, 156, 159, 318, 636, 638] | 15 | [1, 2, 4, 8, 16, 32, 48, 50, 98, 196, 392, 588, 638] | 13 | 104 сек.|

Как можно увидеть, алгоритм дробления вектора индексов тратит гораздо большее время, чем алгоритм Брауэра, но при этом точнее находит минимальную аддитивную цепочку.

*4) Проверка следствия Брауэра*

In [ ]:
import pylab as plt

x = [i for i in range(1, 201)]
ln = [len(splitting_vector(i)) for i in range(1, 201)]
lnx = [math.log2(i) for i in range(1, 201)]
lnx2 = [math.log2(i) ** 2 for i in range(1, 201)]
lnx_y = [ln[i - 1] * lnx[i - 1] for i in range(1, 201)]

a = (201 * sum(lnx_y) - sum(ln) * sum(lnx)) / (201 * sum(lnx2) - sum(lnx) ** 2)
b = (sum(ln) - a * sum(lnx)) / 200
y = [b + a * lnx[i - 1] for i in range(1, 201)]
plt.figure(figsize=(12, 8))
plt.scatter(x, ln, color='r', marker='.')
plt.plot(x, y)
plt.show()

Из графика видно, что функция $l(n)$ аппроксимируется логарифмом, что и является следствием теоремы Брауэра.

*5) Проверка гипотезы Шольца–Брауэра*

Проверим гипотезу Шольца-Брауэра для $n: 1 < n \leq 10$ 

In [ ]:
for i in range(2, 11):
    print(f"n = {i}, l(2^n-1) = {len(splitting_vector(2 ** i - 1))}, l(n)+n-1 = {len(splitting_vector(i)) + i - 1}")

Сведем результаты в таблицу:

| n | l(2^n-1)| l(n)+n-1 |
| :-: | :-: | :-: |
| 2 | 3 | 3 |
| 3 | 5 | 5 |
| 4 | 6 | 6 |
| 5 | 8 | 8 |
| 6 | 9 | 9 |
| 7 | 11 | 11 |
| 8 | 11 | 11 |
| 9 | 13 | 13 |
| 10 | 14 | 14 |

Как можно увидеть, для $2 \leq n \leq 10$ гипотеза выполняется. Также можно заметить, что значения совпали, следовательно при $2 \leq n \leq 10$ для посчета $l(2^n-1)$ можно использовать формулу $l(n)+n-1$, тем самым сэкономив время.

## Вывод

В ходе выполнения практической работы были изучены и реализованы алгоритмы быстрого возведения в степень: бинарный метод и метод множителей. По итогам тестирования было выявлено, что оба алгоритма тратят примерно одинаковое количество операций, но в некоторых ситуациях метод множетелей работал лучше.

Еще было изучено понятие аддитивной цепочки и изучены алгоритмы их построения - алгоритм Брауэра и алгоритм дробления вектора индексов. Алгоритм Брауэра работает гораздно быстрее, но реже находит минимальную аддитивную цепочку, алгоритм ДВИ работает очень долго, но всегда находит минимальную. 

Также было проверено следствие теоремы Брауэра и гипотеза Штольца-Брауэра.